com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.MixedReality.Toolkit.Input.Simulation;
+using UnityEngine;
+using UnityEngine.InputSystem.Utilities;
+using UnityEngine.XR;
+using UnityEngine.XR.Interaction.Toolkit;
+using UnityInputSystem = UnityEngine.InputSystem;
+using InputAction = UnityEngine.InputSystem.InputAction;
+using InputActionProperty = UnityEngine.InputSystem.InputActionProperty;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Basic controller visualizer which renders the a controller model when one is detected.
+    /// The platform controller model is used when available, otherwise a generic controller model is used.
+    /// </summary>
+    [AddComponentMenu("MRTK/Input/Controller Visualizer")]
+    public class ControllerVisualizer : MonoBehaviour
+    {
+        [SerializeField]
+        [Tooltip("The XRNode on which this hand is located.")]
+        private XRNode handNode = XRNode.LeftHand;
+
+        /// <summary> The XRNode on which this hand is located. </summary>
+        public XRNode HandNode { get => handNode; set => handNode = value; }
+
+        [SerializeField]
+        [Tooltip("A fallback controller model to render in case the platform model fails to load")]
+        private GameObject fallbackControllerModel;
+
+        // caching the controller we belong to
+        private XRBaseController xrController;
+
+        // The controller usages we want the input device to have;
+        private InternedString targetUsage;
+
+        // A gameobject representing the currently loaded platform models.
+        private GameObject platformLoadedGameObject;
+
+        // A gameobject representing the root which contains any loaded platform models.
+        // This root is necessary since platform models are rotated 180 degrees by default.
+        private GameObject platformLoadedGameObjectRoot;
```

Comment:
I think these should be seperate because hypothetically, if 1 user was to switch the type of controller they were using on one hand, disabling the root but not the underlying base controller model would result in 2 different kinds of models being overlaid on top of each other. Because we already need to parent the platform loaded model to a MRTK managed root, it's better to explicitly enable/disable that model rather than just rely on enabling/disabling the root.

Reply:
> would result in 2 different kinds of models being overlaid on top of each other.

Interesting...so in this case, we'd load a new model, store it in `platformLoadedGameObject`, overwriting the old model? Do we still have a reference to that in case the original controller is reconnected, or would we just leave it stale in the scene and load a duplicate?

Reply:
Oh I see, that's largely handled by `ControllerModelLoader`